cryptographic_flows.md

```diff
@@ -1,27 +1,24 @@
 
-# Authenticate
-The local client inputs a user ID and a password and receives an "export key", which we will consider as a valid symmetric encryption key, as output. The key server receives as output some user-specific authentication material, which it can compare to its stored data to verify authentication.
-
-Both parties recieve an open channel as output, which satisfies mutual authentication, confidentiality, and integrity.
-
-
 # Generate a secret
 
-The local client component can generate a secret. Using a cryptographically secure pseudo-random number generator with a random seed, generate 256 bits of output.
-
-Generate a unique tag to associate with the secret (unique among secrets stored by the local client).
+The local client component can generate a secret.
+1. Generate the secret: using a cryptographically secure pseudo-random number generator with a random seed, generate 256 bits of output.
+2. Generate a unique tag to associate with the secret (unique among secrets stored by the local client). A tag is 16 bytes (128 bits) of arbitrary data. It can be generated deterministically (e.g. incrementing by 1 for each new secret) or pseudorandomly.
+3. Output the tag and the secret.
 
 # Generate and store a secret locally
-The local client component can run the [generate](generate-a-secret) protocol. Then, store the tag and the secret as a pair.
+The local client component can run this operation.
+1. Execute the [generate](#generate-a-secret) protocol.
+2. Store the tag and the secret as a pair.
 
 # Generate and store a secret remotely
-The local client must
-1. Run the [authenticate](authenticate) protocol to retrieve an export key and an open channel.
-2. Run the [generate](generate-a-secret) protocol to get a tag and a secret.
-3. Use the export key to encrypt the secret.
-4. Send the tag and encrypted secret over the open channel.
+To store a secret remotely, the local client must already have opened a mutually authenticated session with the key server.
 
-The key server generates a unique secret ID for the key by generating 256 bits of randomness and hashing it with the tag to get an ID for the secret (if the ID is not unique among the key server's stored IDs, try again).
-Then, store the user ID (from authentication), the secret ID and encrypted secret.
+TODO (#4 #10 #27): Describe the authentication flow and outputs precisely. The outputs should include a symmetric encryption key and a user ID.
```

Comment:
1. The outputs for whom? 
1. Which symmetric encryption key are you referencing here? Are you saying this is the same key that will be used in L20? 
1. Are there reasons why we might consider having a layer of indirection here? i.e.,
    1. Generate a separate symmetric key that is used to encrypt our arbitrary secrets data, `backup_key`.
    1. Encrypt the `backup_key` with a symmetric key derived from the authentication flow. 
1. Does your answer to the previous question change if you think about use cases where you might want to store your arbitrary secret data _both_ locally and at the server?